lib/src/test/java/io/ably/lib/test/rest/HttpTest.java

```diff
@@ -147,6 +163,118 @@ public Http setUrlArgumentStack(List<String> urlArgumentStack) {
+
+		try {
+			http.ablyHttpExecute(
+					"/path/to/fallback", /* Ignore path */
+					Http.GET, /* Ignore method */
+					new Param[0], /* Ignore headers */
+					new Param[0], /* Ignore params */
+					null, /* Ignore requestBody */
+					null /* Ignore requestHandler */
+			);
+		} catch (AblyException e) {
+			ErrorInfo expectedErrorInfo = new ErrorInfo("Connection failed; no host available", 404, 80000);
+			assertThat(e, new ErrorInfoMatcher(expectedErrorInfo));
+		}
+
+		/* Validate that only one host is used and it is Defaults#HOST_REST */
+		Assert.assertTrue(urlHostArgumentStack.size() == 1);
+		assertThat(urlHostArgumentStack.get(0), is(equalTo(Defaults.HOST_REST)));
+	}
+
+	/**
+	 * Validates that Http uses ClientOptions#fallbackHosts when there is AblyException.HostFailedException thrown
+	 * @throws AblyException
+	 */
+
+	@Test
+	public void http_ably_execute_custom_fallback() throws AblyException {
+		ClientOptions options = new ClientOptions();
+		options.tls = false;
+		options.fallbackHosts = CUSTOM_HOSTS;
+		options.port = 7777;
+
+		ArrayList<String> urlHostArgumentStack = new ArrayList<>();
+
+		Http http = new Http(options, null) {
+			/* Store only string representations to avoid try/catch blocks */
+			List<String> urlArgumentStack;
+
+			@Override
+			public <T> T httpExecuteWithRetry(URL url, String method, Param[] headers, RequestBody requestBody, ResponseHandler<T> responseHandler, boolean allowAblyAuth) throws AblyException {
+				urlArgumentStack.add(url.getHost());
+				/* verify if fallback hosts are from specified list */
+				if(!url.getHost().equals(Defaults.HOST_REST))
+					assertTrue(CUSTOM_HOSTS.contains(url.getHost()));
+
+				return super.httpExecuteWithRetry(url, method, headers, requestBody, responseHandler, allowAblyAuth);
+			}
+
+			public Http setUrlArgumentStack(List<String> urlArgumentStack) {
+				this.urlArgumentStack = urlArgumentStack;
+				return this;
+			}
+		}.setUrlArgumentStack(urlHostArgumentStack);
+
+		http.setHost(Defaults.HOST_REST);
+
+		try {
+			http.ablyHttpExecute(
+					"/path/to/fallback", /* Ignore path */
+					Http.GET, /* Ignore method */
+					new Param[0], /* Ignore headers */
+					new Param[0], /* Ignore params */
+					null, /* Ignore requestBody */
+					null /* Ignore requestHandler */
+			);
+		} catch (AblyException e) {
+			/* Verify that,
+			 * 		- an {@code AblyException} with {@code ErrorInfo} having a `404` status code is thrown.
+			 */
+			ErrorInfo expectedErrorInfo = new ErrorInfo("Connection failed; no host available", 404, 80000);
+			assertThat(e, new ErrorInfoMatcher(expectedErrorInfo));
+		}
+
+		int expectedCallCount = options.httpMaxRetryCount + 1;
+		Assert.assertTrue(urlHostArgumentStack.size() == expectedCallCount);
+		assertThat(urlHostArgumentStack.get(0), is(equalTo(Defaults.HOST_REST)));
+
+		for (int i = 1; i < expectedCallCount; i++) {
+			Assert.assertTrue(urlHostArgumentStack.get(i).matches(PATTERN_HOST_FALLBACK));
```

Comment:
Given this regex is `"(?i)[a-z]\\.ably-realtime.com"`, then if the default fallback hosts are used the this test would still pass?  Shouldn't this be testing that the fallback hosts are actually used?